File: combinatorics/resource/TMD/graph_tree/Code/problem06/python/dynamicprogramming.py
import sys
from typing import List, Dict, Optional, Tuple

# Increase Python's default recursion limit for potentially deep trees.
# The Zhang-Shasha algorithm is heavily recursive.
sys.setrecursionlimit(2000)

class Node:
    """
    Represents a node in a tree with a string label and a list of children nodes.
    """
    def __init__(self, label: str):
        self.label = label
        self.children: List['Node'] = []

    def __repr__(self) -> str:
        # A representation to help with debugging and printing.
        return f"Node(label='{self.label}')"

class DpSolver:
    """
    Solves the Tree Edit Distance problem using Dynamic Programming with memoization.
    This is an efficient implementation of the classic Zhang-Shasha algorithm,
    which can be seen as an optimized version of the Divide and Conquer approach.
    """

    def __init__(self):
        """
        Initializes the solver. The memoization table is the key component for the DP approach.
        It stores the computed distances for pairs of subtrees to avoid redundant calculations.
        """
        self.memo: Dict[Tuple[Optional[Node], Optional[Node]], int] = {}

    def get_subtree_size(self, node: Optional[Node]) -> int:
        """
        Helper function to calculate the total number of nodes in a subtree.
        This is used to determine the cost of inserting or deleting an entire subtree.
        """
        if not node:
            return 0
        # The size is 1 (for the node itself) plus the sizes of all its children's subtrees.
        return 1 + sum(self.get_subtree_size(child) for child in node.children)

    def solve(self, t1: Optional[Node], t2: Optional[Node]) -> int:
        """
        The main recursive function with memoization for calculating TED.
        This is the public entry point for the algorithm. For repeated calls on different
        trees, a new DpSolver instance should be created to clear the memoization table.
        """
        # Step 1: Check the memoization table first.
        # If the result for this pair of subtrees has already been computed, return it instantly.
        if (t1, t2) in self.memo:
            return self.memo[(t1, t2)]
            
        # --- Base Cases ---
        # If one tree is empty, the cost is the number of nodes in the other tree (all insertions/deletions).
        if t1 is None:
            res = self.get_subtree_size(t2)
        elif t2 is None:
            res = self.get_subtree_size(t1)
        else:
            # --- Recursive Step (Divide and Conquer with Combination) ---
            # Option 1: Relabel/substitute the roots of t1 and t2.
            # The cost is the relabeling cost plus the minimum cost to transform the children forests.
            cost_relabel = 0 if t1.label == t2.label else 1
            res_sub = cost_relabel + self._forest_dist_dp(t1.children, t2.children)
            
            # Option 2: Delete the root of t1 (and its entire subtree).
            res_del = self.get_subtree_size(t1)
            
            # Option 3: Insert the root of t2 (and its entire subtree).
            res_ins = self.get_subtree_size(t2)
            
            # The final result for this subproblem is the minimum of the three options.
            res = min(res_sub, res_del, res_ins)
        
        # Step 2: Store the newly computed result in the memoization table before returning.
        self.memo[(t1, t2)] = res
        return res

    def _forest_dist_dp(self, f1: List[Node], f2: List[Node]) -> int:
        """
        Calculates the edit distance between two forests (sequences of trees)
        using a dynamic programming table, similar to the Levenshtein distance for strings.
        """
        n, m = len(f1), len(f2)
        # Create a DP table to solve the subproblem.
        dp_table = [[0] * (m + 1) for _ in range(n + 1)]

        # Base cases for the DP table: transforming a forest to an empty one (and vice versa).
        # Cost is the sum of sizes of all trees in the forest.
        for i in range(1, n + 1):
            dp_table[i][0] = dp_table[i-1][0] + self.get_subtree_size(f1[i-1])
        for j in range(1, m + 1):
            dp_table[0][j] = dp_table[0][j-1] + self.get_subtree_size(f2[j-1])
            
        # Fill the rest of the table using the main recursive `solve` function.
        # Since `solve` is memoized, calls for the same subtrees are fast.
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                # Cost to transform tree i of f1 into tree j of f2.
                cost_sub = self.solve(f1[i-1], f2[j-1])
                # Cost to delete tree i of f1.
                cost_del = self.get_subtree_size(f1[i-1])
                # Cost to insert tree j of f2.
                cost_ins = self.get_subtree_size(f2[j-1])
                
                dp_table[i][j] = min(
                    dp_table[i-1][j-1] + cost_sub, # Substitute
                    dp_table[i-1][j] + cost_del,     # Deletion
                    dp_table[i][j-1] + cost_ins      # Insertion
                )
        return dp_table[n][m]

def build_from_parents(parents: List[int], labels: List[str]) -> Optional[Node]:
    """Helper function to build a tree from a parent array for testing."""
    if not parents:
        return None
    nodes = [Node(lbl) for lbl in labels]
    root = None
    for i, p_idx in enumerate(parents):
        if p_idx == -1:
            root = nodes[i]
        else:
            nodes[p_idx].children.append(nodes[i])
    return root

# =================================================================================
# MAIN EXECUTION BLOCK - DEMONSTRATION FOR EXERCISE 6(d)
# =================================================================================
if __name__ == '__main__':
    print("=" * 60)
    print("  (d) Tree Edit Distance: Dynamic Programming Solver  ")
    print("=" * 60)
    
    # --- Test Cases ---
    
    # Test Case 1: Simple relabel
    # Tree 1: a -> {b, c}
    t1 = build_from_parents([-1, 0, 0], ["a", "b", "c"])
    # Tree 2: a -> {d, c}
    t2 = build_from_parents([-1, 0, 0], ["a", "d", "c"])
    solver1 = DpSolver()
    print(f"Distance( a{{b,c}}, a{{d,c}} ) = {solver1.solve(t1, t2)} (Expected: 1)")

    # Test Case 2: More complex transformation
    # Tree 3: x -> {y, z}
    t3 = build_from_parents([-1, 0, 0], ["x", "y", "z"])
    solver2 = DpSolver()
    print(f"Distance( a{{b,c}}, x{{y,z}} ) = {solver2.solve(t1, t3)} (Expected: 3)")

    # Test Case 3: Different structures
    # Tree 4: f -> {a -> {b, c}, d -> {e}}
    t4 = build_from_parents([-1, 0, 0, 1, 1, 2], ["f", "a", "d", "b", "c", "e"])
    # Tree 5: f -> {c, d -> {a, b, e}}
    t5 = build_from_parents([-1, 0, 0, 2, 2, 2], ["f", "c", "d", "a", "b", "e"])
    solver3 = DpSolver()
    print(f"Distance( T4, T5 ) = {solver3.solve(t4, t5)} (Expected: 2)")

    print("=" * 60)

File: combinatorics/resource/TMD/graph_tree/Code/problem06/c++/dynamicprogramming.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <memory>
#include <unordered_map>

// --- Data Structures ---
// Using raw pointers for this version to simplify map keys.
// Memory management must be done manually.
struct Node {
    std::string label;
    std::vector<Node*> children;
    Node(std::string lbl) : label(std::move(lbl)) {}
    ~Node() {
        for(Node* child : children) {
            delete child;
        }
    }
};

// Custom hash function for pairs of Node pointers, required for unordered_map
struct NodePtrPairHash {
    std::size_t operator()(const std::pair<const Node*, const Node*>& p) const {
        auto h1 = std::hash<const Node*>{}(p.first);
        auto h2 = std::hash<const Node*>{}(p.second);
        // Combine hashes to reduce collisions
        return h1 ^ (h2 << 1);
    }
};

class DpSolver {
private:
    // Memoization table: stores the computed distance for a pair of subtrees.
    // This is the core of the dynamic programming optimization.
    std::unordered_map<std::pair<const Node*, const Node*>, int, NodePtrPairHash> memo;

public:
    /**
     * @brief Public entry point for the Dynamic Programming solver.
     * This method clears any previous memoization results and starts the recursion.
     * @param t1 The root of the first tree.
     * @param t2 The root of the second tree.
     * @return The minimum tree edit distance.
     */
    int solve(const Node* t1, const Node* t2) {
        memo.clear();
        return ted_dp(t1, t2);
    }

private:
    /**
     * @brief Helper to get the total number of nodes in a subtree.
     */
    int get_subtree_size(const Node* node) {
        if (!node) return 0;
        int size = 1;
        for (const auto* child : node->children) {
            size += get_subtree_size(child);
        }
        return size;
    }

    /**
     * @brief The core recursive function with memoization for calculating TED.
     *
     * This function implements the Zhang-Shasha algorithm. It's a divide-and-conquer
     * approach enhanced with a memoization table (`memo`) to store and reuse results
     * of overlapping subproblems, changing the complexity from exponential to polynomial.
     */
    int ted_dp(const Node* t1, const Node* t2) {
        // Step 1: Check memoization table first. If result exists, return it immediately.
        if (memo.count({t1, t2})) {
            return memo.at({t1, t2});
        }
            
        int res;
        // Base cases: one of the subtrees is empty.
        if (!t1) {
            res = get_subtree_size(t2);
        } else if (!t2) {
            res = get_subtree_size(t1);
        } else {
            // Recursive step: calculate costs for the three main operations.
            int cost_relabel = (t1->label == t2->label) ? 0 : 1;
            int res_sub = cost_relabel + forest_dist_dp(t1->children, t2->children);
            
            // Note: The classic Zhang-Shasha definition is used for delete/insert costs
            // which involves mapping a tree to an empty node or vice-versa.
            int res_del = 1 + ted_dp(nullptr, t2) + forest_dist_dp(t1->children, {}); // Cost to delete t1's root
            int res_ins = 1 + ted_dp(t1, nullptr) + forest_dist_dp({}, t2->children); // Cost to insert t2's root
            
            res = std::min({res_sub, res_del, res_ins});
        }
        
        // Step 2: Store the newly computed result in the memoization table before returning.
        return memo[{t1, t2}] = res;
    }

    /**
     * @brief Calculates the edit distance between two forests using a DP table.
     * This is a subproblem within the main TED calculation.
     */
    int forest_dist_dp(const std::vector<Node*>& f1, const std::vector<Node*>& f2) {
        int n = f1.size();
        int m = f2.size();
        std::vector<std::vector<int>> dp_table(n + 1, std::vector<int>(m + 1, 0));

        // Base cases for the DP table (mapping a forest to an empty one).
        for (int i = 1; i <= n; ++i) 
            dp_table[i][0] = dp_table[i-1][0] + get_subtree_size(f1[i-1]);
        for (int j = 1; j <= m; ++j) 
            dp_table[0][j] = dp_table[0][j-1] + get_subtree_size(f2[j-1]);

        // Fill the rest of the table using recursive calls to the memoized `ted_dp`.
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                int cost_sub = ted_dp(f1[i-1], f2[j-1]);
                int cost_del = get_subtree_size(f1[i-1]);
                int cost_ins = get_subtree_size(f2[j-1]);
                
                dp_table[i][j] = std::min({dp_table[i-1][j-1] + cost_sub,
                                           dp_table[i-1][j] + cost_del,
                                           dp_table[i][j-1] + cost_ins});
            }
        }
        return dp_table[n][m];
    }
};

/**
 * @brief Helper function to build a tree from a parent array for testing.
 */
Node* build_from_parents(const std::vector<int>& parents, const std::vector<std::string>& labels) {
    int n = parents.size();
    if (n == 0) return nullptr;
    
    std::vector<Node*> nodes;
    for(int i = 0; i < n; ++i) {
        nodes.push_back(new Node(labels[i]));
    }
    
    Node* root = nullptr;
    for(int i = 0; i < n; ++i) {
        if(parents[i] != -1) {
            nodes[parents[i]]->children.push_back(nodes[i]);
        } else {
            root = nodes[i];
        }
    }
    return root;
}

// =================================================================================
// MAIN FUNCTION - DEMONSTRATION FOR EXERCISE 6(d)
// =================================================================================
int main() {
    std::cout << "=====================================================" << std::endl;
    std::cout << "  (d) Tree Edit Distance: Dynamic Programming        " << std::endl;
    std::cout << "=====================================================" << std::endl;
    
    DpSolver solver;
    
    // --- Test Cases ---
    
    // Test Case 1: Simple relabel
    Node* t1 = build_from_parents({-1, 0, 0}, {"a", "b", "c"});
    Node* t2 = build_from_parents({-1, 0, 0}, {"a", "d", "c"});
    std::cout << "Distance( a{b,c}, a{d,c} ) = " << solver.solve(t1, t2) << " (Expected: 1)" << std::endl;

    // Test Case 2: More complex transformation
    Node* t3 = build_from_parents({-1, 0, 2}, {"x", "y", "z"});
    std::cout << "Distance( a{b,c}, x{y,z{}} ) = " << solver.solve(t1, t3) << " (Expected: 3)" << std::endl;

    // Test Case 3: Larger trees
    Node* t4 = build_from_parents({-1, 0, 0, 1, 1, 2}, {"f", "a", "d", "b", "c", "e"});
    Node* t5 = build_from_parents({-1, 0, 0, 2, 2, 2}, {"f", "c", "d", "a", "b", "e"});
    std::cout << "Distance( f{a{b,c},d{e}}, f{c,d{a,b,e}} ) = " << solver.solve(t4, t5) << " (Expected: 2)" << std::endl;

    // --- Cleanup ---
    delete t1;
    delete t2;
    delete t3;
    delete t4;
    delete t5;

    return 0;
}